File: luau-obfuscator/src/lexer/lexer.ts
/**
 * Lexer: converts Luau source into a flat token stream. Byte-oriented (operates
 * on char codes) so it round-trips arbitrary string contents. Records the exact
 * `raw` text of every token so the generator can reproduce literals faithfully.
 */
import { lexError, type SourceLocation } from '../diagnostics.js';
import { KEYWORDS, SYMBOLS, type Token } from './token.js';

function isDigit(c: string): boolean { return c >= '0' && c <= '9'; }
function isHex(c: string): boolean {
  return (c >= '0' && c <= '9') || (c >= 'a' && c <= 'f') || (c >= 'A' && c <= 'F');
}
function isAlpha(c: string): boolean {
  return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || c === '_';
}
function isAlphaNum(c: string): boolean { return isAlpha(c) || isDigit(c); }

export function tokenize(src: string, chunk = 'input'): Token[] {
  const tokens: Token[] = [];
  const n = src.length;
  let pos = 0;
  let line = 1;

  const loc = (): SourceLocation => ({ chunk, line });
  const peek = (o = 0): string => src.charAt(pos + o);

  function newline(): void {
    const c = peek();
    pos++;
    const c2 = peek();
    if ((c2 === '\n' || c2 === '\r') && c2 !== c) pos++;
    line++;
  }

  function readLongBracket(): string | null {
    if (peek() !== '[') return null;
    let level = 0;
    let p = pos + 1;
    while (src.charAt(p) === '=') { level++; p++; }
    if (src.charAt(p) !== '[') return null;
    pos = p + 1;
    if (peek() === '\r' || peek() === '\n') newline();
    const buf: string[] = [];
    const close = ']' + '='.repeat(level) + ']';
    for (;;) {
      if (pos >= n) lexError('unfinished long bracket', loc());
      const c = peek();
      if (c === ']' && src.substr(pos, close.length) === close) {
        pos += close.length;
        return buf.join('');
      } else if (c === '\n' || c === '\r') {
        buf.push('\n'); newline();
      } else { buf.push(c); pos++; }
    }
  }

  function readString(quote: string): string {
    pos++;
    const buf: string[] = [];
    for (;;) {
      if (pos >= n) lexError('unfinished string', loc());
      const c = peek();
      if (c === quote) { pos++; break; }
      if (c === '\n' || c === '\r') lexError('unfinished string', loc());
      if (c === '\\') {
        pos++;
        const e = peek();
        switch (e) {
          case 'n': buf.push('\n'); pos++; break;
          case 't': buf.push('\t'); pos++; break;
          case 'r': buf.push('\r'); pos++; break;
          case 'a': buf.push('\x07'); pos++; break;
          case 'b': buf.push('\b'); pos++; break;
          case 'f': buf.push('\f'); pos++; break;
          case 'v': buf.push('\x0b'); pos++; break;
          case '\\': buf.push('\\'); pos++; break;
          case '"': buf.push('"'); pos++; break;
          case "'": buf.push("'"); pos++; break;
          case '\n': case '\r': buf.push('\n'); newline(); break;
          case 'x': {
            pos++;
            let h = '';
            for (let k = 0; k < 2; k++) { if (isHex(peek())) { h += peek(); pos++; } }
            if (h.length === 0) lexError('hexadecimal digit expected', loc());
            buf.push(String.fromCharCode(parseInt(h, 16)));
            break;
          }
          case 'z': {
            pos++;
            while (pos < n) {
              const w = peek();
              if (w === '\n' || w === '\r') newline();
              else if (w === ' ' || w === '\t' || w === '\f' || w === '\x0b') pos++;
              else break;
            }
            break;
          }
          default: {
            if (isDigit(e)) {
              let d = '';
              for (let j = 0; j < 3; j++) { if (isDigit(peek())) { d += peek(); pos++; } else break; }
              const num = parseInt(d, 10);
              if (num > 255) lexError('decimal escape too large', loc());
              buf.push(String.fromCharCode(num));
            } else {
              lexError(`invalid escape sequence '\\${e}'`, loc());
            }
          }
        }
      } else { buf.push(c); pos++; }
    }
    return buf.join('');
  }

  function readNumber(): string {
    const start = pos;
    if (peek() === '0' && (peek(1) === 'x' || peek(1) === 'X')) {
      pos += 2;
      while (isHex(peek()) || peek() === '.' || peek() === '_') pos++;
      if (peek() === 'p' || peek() === 'P') {
        pos++;
        if (peek() === '+' || peek() === '-') pos++;
        while (isDigit(peek())) pos++;
      }
    } else {
      // decimal / binary (Luau 0b) with digit separators
      if (peek() === '0' && (peek(1) === 'b' || peek(1) === 'B')) {
        pos += 2;
        while (peek() === '0' || peek() === '1' || peek() === '_') pos++;
      } else {
        while (isDigit(peek()) || peek() === '.' || peek() === '_') pos++;
        if (peek() === 'e' || peek() === 'E') {
          pos++;
          if (peek() === '+' || peek() === '-') pos++;
          while (isDigit(peek())) pos++;
        }
      }
    }
    return src.substring(start, pos);
  }

  while (pos < n) {
    const c = peek();
    const tokStart = pos;
    if (c === '\n' || c === '\r') {
      newline();
    } else if (c === ' ' || c === '\t' || c === '\f' || c === '\x0b') {
      pos++;
    } else if (c === '-' && peek(1) === '-') {
      pos += 2;
      if (peek() === '[') {
        const saved = pos;
        const content = readLongBracket();
        if (content === null) {
          pos = saved;
          while (pos < n && peek() !== '\n' && peek() !== '\r') pos++;
        }
      } else {
        while (pos < n && peek() !== '\n' && peek() !== '\r') pos++;
      }
    } else if (isAlpha(c)) {
      while (isAlphaNum(peek())) pos++;
      const word = src.substring(tokStart, pos);
      tokens.push({ type: KEYWORDS.has(word) ? 'keyword' : 'name', value: word, line, raw: word });
    } else if (isDigit(c) || (c === '.' && isDigit(peek(1)))) {
      const ln = line;
      const numText = readNumber();
      tokens.push({ type: 'number', value: numText, line: ln, raw: numText });
    } else if (c === '"' || c === "'") {
      const ln = line;
      const sval = readString(c);
      tokens.push({ type: 'string', value: sval, line: ln, raw: src.substring(tokStart, pos), long: false });
    } else if (c === '[' && (peek(1) === '[' || peek(1) === '=')) {
      const ln = line;
      const lc = readLongBracket();
      if (lc !== null) {
        tokens.push({ type: 'string', value: lc, line: ln, raw: src.substring(tokStart, pos), long: true });
      } else {
        tokens.push({ type: 'symbol', value: '[', line: ln, raw: '[' });
        pos++;
      }
    } else {
      let matched: string | null = null;
      for (const sym of SYMBOLS) {
        if (src.substr(pos, sym.length) === sym) { matched = sym; break; }
      }
      if (matched === null) lexError(`unexpected symbol near '${c}'`, loc());
      tokens.push({ type: 'symbol', value: matched, line, raw: matched });
      pos += matched.length;
    }
  }

  tokens.push({ type: 'eof', value: '<eof>', line, raw: '' });
  return tokens;
}

File: capybara/lexer.lua
-- capybara/lexer.lua
-- A standalone tokenizer for Lua 5.x / Luau source. Emits a flat token list:
--   { type = "name"|"keyword"|"number"|"string"|"symbol"|"eof",
--     value = <lua value or text>, raw = <exact source text>, line = <int> }
--
-- Tokenizing Lua is a well-known, generic problem; this is our own
-- implementation written for capybara. It keeps `raw` for every token so the
-- emitter can re-serialize losslessly, and decodes string/number values so the
-- transform layers can operate on real values.

local Lexer = {}

local KEYWORDS = {
    ["and"] = true, ["break"] = true, ["do"] = true, ["else"] = true,
    ["elseif"] = true, ["end"] = true, ["false"] = true, ["for"] = true,
    ["function"] = true, ["goto"] = true, ["if"] = true, ["in"] = true,
    ["local"] = true, ["nil"] = true, ["not"] = true, ["or"] = true,
    ["repeat"] = true, ["return"] = true, ["then"] = true, ["true"] = true,
    ["until"] = true, ["while"] = true, ["continue"] = true,
}

-- Longest-match first so greedy scanning picks the right operator.
local SYMBOLS = {
    "...", "..", "::", "==", "~=", "<=", ">=", "//", "<<", ">>",
    "+=", "-=", "*=", "/=", "%=", "^=", "..=", -- Luau compound-assign
    "+", "-", "*", "/", "%", "^", "#", "&", "~", "|", "<", ">", "=",
    "(", ")", "{", "}", "[", "]", ";", ":", ",", ".",
}
-- ".." variants must be tested before "." but "..=" before ".." — resort by length.
table.sort(SYMBOLS, function(a, b) return #a > #b end)

local function isDigit(c) return c >= "0" and c <= "9" end
local function isHex(c)
    return (c >= "0" and c <= "9") or (c >= "a" and c <= "f") or (c >= "A" and c <= "F")
end
local function isAlpha(c)
    return (c >= "a" and c <= "z") or (c >= "A" and c <= "Z") or c == "_"
end
local function isAlphaNum(c) return isAlpha(c) or isDigit(c) end

function Lexer.tokenize(src, chunkname)
    chunkname = chunkname or "input"
    -- skip a leading shebang line if present
    if src:sub(1, 1) == "#" then
        local nl = src:find("\n") or (#src + 1)
        src = src:sub(nl)
    end

    local tokens = {}
    local pos, line = 1, 1
    local n = #src

    local function err(msg)
        error(string.format("%s:%d: %s", chunkname, line, msg), 0)
    end
    local function peek(o) return src:sub(pos + (o or 0), pos + (o or 0)) end

    local function newline()
        local c = peek()
        pos = pos + 1
        local c2 = peek()
        if (c2 == "\n" or c2 == "\r") and c2 ~= c then pos = pos + 1 end
        line = line + 1
    end

    -- long bracket [[ ]] / [==[ ]==]; returns content or nil when not one
    local function readLongBracket()
        if peek() ~= "[" then return nil end
        local level, p = 0, pos + 1
        while src:sub(p, p) == "=" do level = level + 1; p = p + 1 end
        if src:sub(p, p) ~= "[" then return nil end
        pos = p + 1
        if peek() == "\r" or peek() == "\n" then newline() end
        local buf = {}
        local close = "]" .. string.rep("=", level) .. "]"
        while true do
            if pos > n then err("unfinished long bracket") end
            local c = peek()
            if c == "]" and src:sub(pos, pos + #close - 1) == close then
                pos = pos + #close
                return table.concat(buf)
            elseif c == "\n" or c == "\r" then
                buf[#buf + 1] = "\n"; newline()
            else
                buf[#buf + 1] = c; pos = pos + 1
            end
        end
    end

    local function readString(quote)
        pos = pos + 1
        local buf = {}
        while true do
            if pos > n then err("unfinished string") end
            local c = peek()
            if c == quote then
                pos = pos + 1; break
            elseif c == "\n" or c == "\r" then
                err("unfinished string")
            elseif c == "\\" then
                pos = pos + 1
                local e = peek()
                local simple = { n = "\n", t = "\t", r = "\r", a = "\a", b = "\b",
                                 f = "\f", v = "\v", ["\\"] = "\\", ['"'] = '"', ["'"] = "'" }
                if simple[e] then
                    buf[#buf + 1] = simple[e]; pos = pos + 1
                elseif e == "\n" or e == "\r" then
                    buf[#buf + 1] = "\n"; newline()
                elseif e == "x" then
                    pos = pos + 1
                    local h = ""
                    for _ = 1, 2 do if isHex(peek()) then h = h .. peek(); pos = pos + 1 end end
                    if #h == 0 then err("hexadecimal digit expected") end
                    buf[#buf + 1] = string.char(tonumber(h, 16))
                elseif e == "z" then
                    pos = pos + 1
                    while pos <= n do
                        local w = peek()
                        if w == "\n" or w == "\r" then newline()
                        elseif w == " " or w == "\t" or w == "\f" or w == "\v" then pos = pos + 1
                        else break end
                    end
                elseif isDigit(e) then
                    local d = ""
                    for _ = 1, 3 do
                        if isDigit(peek()) then d = d .. peek(); pos = pos + 1 else break end
                    end
                    local num = tonumber(d)
                    if num > 255 then err("decimal escape too large") end
                    buf[#buf + 1] = string.char(num)
                else
                    err("invalid escape sequence '\\" .. e .. "'")
                end
            else
                buf[#buf + 1] = c; pos = pos + 1
            end
        end
        return table.concat(buf)
    end

    local function readNumber()
        local start = pos
        if peek() == "0" and (peek(1) == "x" or peek(1) == "X") then
            pos = pos + 2
            while isHex(peek()) or peek() == "." do pos = pos + 1 end
            if peek() == "p" or peek() == "P" then
                pos = pos + 1
                if peek() == "+" or peek() == "-" then pos = pos + 1 end
                while isDigit(peek()) do pos = pos + 1 end
            end
        else
            while isDigit(peek()) or peek() == "." do pos = pos + 1 end
            if peek() == "e" or peek() == "E" then
                pos = pos + 1
                if peek() == "+" or peek() == "-" then pos = pos + 1 end
                while isDigit(peek()) do pos = pos + 1 end
            end
        end
        local text = src:sub(start, pos - 1)
        local value = tonumber((text:gsub("_", ""))) -- tolerate Luau digit separators
        if value == nil then err("malformed number near '" .. text .. "'") end
        return value, text
    end

    while pos <= n do
        local c = peek()
        local tokStart = pos
        if c == "\n" or c == "\r" then
            newline()
        elseif c == " " or c == "\t" or c == "\f" or c == "\v" then
            pos = pos + 1
        elseif c == "-" and peek(1) == "-" then
            pos = pos + 2
            if peek() == "[" then
                local saved = pos
                if readLongBracket() == nil then
                    pos = saved
                    while pos <= n and peek() ~= "\n" and peek() ~= "\r" do pos = pos + 1 end
                end
            else
                while pos <= n and peek() ~= "\n" and peek() ~= "\r" do pos = pos + 1 end
            end
        elseif isAlpha(c) then
            while isAlphaNum(peek()) do pos = pos + 1 end
            local word = src:sub(tokStart, pos - 1)
            tokens[#tokens + 1] = {
                type = KEYWORDS[word] and "keyword" or "name",
                value = word, raw = word, line = line,
            }
        elseif isDigit(c) or (c == "." and isDigit(peek(1))) then
            local ln = line
            local value = readNumber()
            tokens[#tokens + 1] = { type = "number", value = value, line = ln, raw = src:sub(tokStart, pos - 1) }
        elseif c == '"' or c == "'" then
            local ln = line
            local value = readString(c)
            tokens[#tokens + 1] = { type = "string", value = value, line = ln, raw = src:sub(tokStart, pos - 1), long = false }
        elseif c == "[" and (peek(1) == "[" or peek(1) == "=") then
            local ln = line
            local content = readLongBracket()
            if content ~= nil then
                tokens[#tokens + 1] = { type = "string", value = content, line = ln, raw = src:sub(tokStart, pos - 1), long = true }
            else
                tokens[#tokens + 1] = { type = "symbol", value = "[", line = ln, raw = "[" }
                pos = pos + 1
            end
        else
            local matched
            for _, sym in ipairs(SYMBOLS) do
                if src:sub(pos, pos + #sym - 1) == sym then matched = sym; break end
            end
            if not matched then err("unexpected symbol near '" .. c .. "'") end
            tokens[#tokens + 1] = { type = "symbol", value = matched, line = line, raw = matched }
            pos = pos + #matched
        end
    end

    tokens[#tokens + 1] = { type = "eof", value = "<eof>", line = line }
    return tokens
end

return Lexer
